Activation logic for testing consensus changes

ajtowns opened the pull request:
Buries taproot, and then sets up an alternative activation logic that allows many deployments to be tested concurrently, and for deployments to be de-activated by signet/regtest miners in order to make way for an updated version of soft-forks being tested, while remaining compatible with older node software.

Forward port of #2.

Commit:
Bury taproot deployment

ajtowns committed:
deploymentinfo: report signalling

ajtowns committed:
chainparams: Add support for renouncing an HereticalDeployment

naumenkogs commented on test/functional/feature_config_args.py:
```diff
@@ -45,11 +45,11 @@ def test_config_file_parser(self):
-        with open(inc_conf_file_path, 'w', encoding='utf-8') as conf:
-            conf.write('acceptnonstdtxn=1\n')
-        self.nodes[0].assert_start_raises_init_error(extra_args=[f"-conf={main_conf_file_path}"], expected_msg='Error: acceptnonstdtxn is not currently supported for main chain')
+        #main_conf_file_path = os.path.join(self.options.tmpdir, 'node0', 'bitcoin_main.conf')
```
78b678fb007d93a35d4817f60a9d80b4cb6b2afe
This should be deleted altogether?

ajtowns reviewed (commented):
Review notes:

### bury taproot deployment

We don't want to make taproot a heretical deployment, so bury it.

Compare with bitcoin#23505 . Note that taproot activation has minimal actual effect since bitcoin#23536 and confirm by searching for uses of `DEPLOYMENT_TAPROOT` in the codebase -- it only affects reporting in rpc/blockchain.cpp (and hence there's no value in giving it a special height in `rpc_blockchain.py`). Check activation heights on mainnet/testnet with `bitcoin-cli getdeploymentinfo | jq .deployments.taproot.height`. 

### tests: drop versionbits specific tests

test/versionbits_tests.cpp and tests/fuzz/versionsbits.cpp are removed entirely

Makefile.test.include, ../test/functional/test_runner.py are update to match the removed tests.

### validation: drop "Unknown new rules activated" warning

validation.cpp is updated to drop the version bits warning functionality, as this can't be supported in the same manner via the heretical versionbits logic. Likewise ../test/functional/feature_versionbits_warning.py is removed.

### Replace versionbits for heretical deployments

The meat of the PR. See https://github.com/bitcoin-inquisition/bitcoin/wiki/Heretical-Deployments for the goal.

consensus/params.h replaces `BIP9Deployment` with `HereticalDeployment` and new parameters. 

chainparams.cpp introduces a helper struct `SetupDeployment` so that specifying a heretical deployment is tidier. This is then used to specify heretical parameters for `TESTDUMMY` for each chain. We also reduce the `nMinerConfirmationWindow` for signet from 2016 blocks to 432 blocks so that forks can be activated/deactivated in a few days' worth of blocks rather than a few weeks. We also drop the ability to modify `min_activation_height` on regtest, since that does not exist for heretical deployments.

deploymentinfo.h introduces constants for the activate/abandon flag at the top of versionbits and some functions to convert a bip/bip version into a version number accordingly. (Those functions were used in `SetupDeployment`)

versionbits.h is modified to introduce the updated states compared to BIP9; and drops the stats entirely, since you now only need a single block to trigger activation/abandonment, not some percentage of blocks. `AbstractThresholdConditionChecker` drops the `params` argument entirely (instead `params` will be passed into the constructor), and the methods are updated to match the heretical parameters.

versionbits.cpp is updated correspondingly and implements to heretical activation/abandonment logic. Note that `ACTIVE` is no longer a final state.

deploymentstatus.h is updated so that `DeploymentActiveAfter()` returns true for both `ACTIVE` and `DEACTIVATING`, but false otherwise.

rpc/blockchain.cpp is updated to report on heretical states.

rpc/mining.cpp is updated to only report to GBT when the rule is active, not whether or not signalling is possible.

 ../test/functional/mining_basic.py and rpc_blockchain.py are updated to match the changes to rpc/mining.cpp and rpc/blockchain.cpp.

### tests: add heretical activation fuzz tester

Replace test/fuzz/versionbits.cpp with a new fuzz tester for heretical deployments.

### deploymentinfo: report signalling

Replaces the removed bip9 statistics from `getdeploymentinfo` with a simple array listing blocks in the previous period that signalled for activation/abandonment.

### chainparams: Add support for renouncing an HereticalDeployment

Adds `RenounceDeployments` function which checks ArgsManager for the `-renounce` option, and resets any specified deployments to `NEVER_ACTIVE`. Called after setting up the deployments for both regtest and signet.

naumenkogs commented:
d3028d44d97629f821ea60c62515fd775a790f9b is heavy for review, if you had time it might be useful to split some things out: e.g., alert removal, fuzzing.

naumenkogs commented on src/rpc/blockchain.cpp:
```diff
@@ -1150,50 +1151,48 @@ static void SoftForkDescPushBack(const CBlockIndex* blockindex, UniValue& softfo
+        case ThresholdState::DEFINED:
+        case ThresholdState::STARTED:
+            bip9.pushKV("signal_activate", strprintf("%08x", chainman.GetConsensus().vDeployments[id].signal_activate));
+            [[fallthrough]];
```
Why not using a `break` here? I assumed these states are mutually exclusive.

naumenkogs commented on src/versionbits.cpp:
```diff
@@ -175,6 +175,42 @@ int AbstractThresholdConditionChecker::GetStateSinceHeightFor(const CBlockIndex*
+                } else if ((pindex->nVersion & 0xFF000000l) == VERSIONBITS_TOP_ABANDON) {
+                    s.activate = false;
+                    result.push_back(s);
+                }
```
Is it worth doing something here instead of just dropping `s` on the floor?

ajtowns replied:
Wow, not a fan of the "l" (ell) in that font looking quite so much like a "1" (one).

Maybe it could make sense to test it as:

```c++
const uint8_t maybe_version = static_cast<uint8_t>(pindex->nVersion  & 0xFF);
if ((activate >> 8) == (pindex->nVersion >> 8)) {
     result.push_back({.height = pindex->nHeight, .bip_version = maybe_verison, .activate = true});
} else if ((abandon >> 8) == (pindex->nVersion >> 8)) {
     result.push_back({.height = pindex->nHeight, .bip_version = maybe_verison, .activate = false});
}
```

If the top bits aren't activate or abandon, they might just be using bip-9 versionbits signalling or similar, so I think it's better not to try to guess at what they might mean...

naumenkogs replied:
Now that i'm thinking from the practical angle, i think it's fine either way :)

naumenkogs commented:
utACK 3e8074faa324b75b7c335d38ef0ebb38fca0164f
I reviewed the code and the rationale ([Heretical Deployment](https://github.com/bitcoin-inquisition/bitcoin/wiki/Heretical-Deployments)). It all makes sense.
I haven't reviewed tests and fuzzing.

ajtowns committed:
tests: drop versionbits specific tests

After we replace versionbits with heretical deployments, these will not
make sense.

ajtowns committed:
validation: drop "Unknown new rules activated" warning

ajtowns committed:
Replace versionbits for heretical deployments

ajtowns committed:
tests: add heretical activation fuzz tester

ajtowns commented:
I've split out the big commit a bit as suggested

naumenkogs commented:
utACK 9f46c03
